store/tikv/2pc_exec.go

```diff
@@ -0,0 +1,550 @@
+		zap.Uint64("conn", c.committer.connID),
+		zap.Stringer("action type", c.action),
+		zap.Error(err),
+		zap.Uint64("txnStartTS", c.committer.startTS))
+
+	if c.cancel != nil {
+		old := atomic.SwapInt32(&c.stop, 1)
+		if old == 0 {
+			logutil.BgLogger().Debug("2PC doActionOnBatch to cancel other actions",
+				zap.Uint64("conn", c.committer.connID),
+				zap.Stringer("action type", c.action),
+				zap.Uint64("txnStartTS", c.committer.startTS))
+			c.cancel()
+		}
+	}
+
+	select {
+	case c.errCh <- err:
+	default:
+	}
+}
+
+func (c *execController) terminated() bool {
+	return atomic.LoadInt32(&c.stop) == 1
+}
+
+func (c *execController) updateSingleBatchDetail(bo *Backoffer, beforeSleep int) {
+	detail := c.committer.getDetail()
+	if detail == nil {
+		// lock operations of pessimistic-txn will let commitDetail be nil
+		return
+	}
+	delta := bo.totalSleep - beforeSleep
+	if delta <= 0 {
+		return
+	}
+	atomic.AddInt64(&detail.CommitBackoffTime, int64(delta)*int64(time.Millisecond))
+	detail.Mu.Lock()
+	detail.Mu.BackoffTypes = append(detail.Mu.BackoffTypes, bo.types...)
+	detail.Mu.Unlock()
+}
+
+type mutation struct {
+	key               []byte
+	value             []byte
+	isPessimisticLock bool
+	op                pb.Op
+}
+
+type mutations interface {
+	Iter(start, end []byte) mutationsIter
+	Len() int
+}
+
+type mutationsIter interface {
+	Next() mutation
+	WithFilter(func([]byte) bool)
+}
+
+type mutationWithRegion struct {
+	mutation
+	region RegionVerID
+}
+
+type mapWithRegionIter struct {
+	src mutationsIter
+
+	rc  *RegionCache
+	loc *KeyLocation
+	bo  *Backoffer
+}
+
+func (c *twoPhaseCommitter) mapWithRegion(bo *Backoffer, src mutationsIter) *mapWithRegionIter {
+	return &mapWithRegionIter{
+		src: src,
+		rc:  c.store.regionCache,
+		bo:  bo,
+	}
+}
+
+func (it *mapWithRegionIter) Next() (mutationWithRegion, error) {
+	m := it.src.Next()
+	if m.key == nil {
+		return mutationWithRegion{}, nil
+	}
+
+	var err error
+	if it.loc == nil || !it.loc.Contains(m.key) {
+		it.loc, err = it.rc.LocateKey(it.bo, m.key)
+		if err != nil {
+			return mutationWithRegion{}, errors.Trace(err)
+		}
+	}
+
+	return mutationWithRegion{m, it.loc.Region}, nil
+}
+
+type mutationBatchCollector struct {
+	src        *mapWithRegionIter
+	sizer      func(*mutation) int
+	primaryKey []byte
+	limit      int
+	init       bool
+	done       bool
+	collector  func(*CommitterMutations, *mutation)
+
+	curr    mutationWithRegion
+	lenHint int
+}
+
+func (c *mutationBatchCollector) Collect() (*batchMutations, error) {
+	if !c.init {
+		m, err := c.src.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		c.curr = m
+		c.init = true
+	}
+	if c.done {
+		return nil, nil
+	}
+
+	var (
+		mutations = NewCommiterMutations(c.lenHint)
+		region    = c.curr.region
+		m         = c.curr
+		isPrimary bool
+		size      int
+		err       error
+	)
+
+	for {
+		if m.key == nil {
+			c.done = true
+			break
+		}
+
+		c.collector(&mutations, &m.mutation)
```

Comment:
Can we use a bool `keyOnly` instead of an abstract function?

Reply:
A `collector` may be more expressive and easier to extend.

Reply:
Is there any need to extend it?